pandas/core/indexes/interval.py

```diff
@@ -1077,28 +1114,68 @@ def equals(self, other):
     def overlaps(self, other):
         return self._data.overlaps(other)
 
-    def _setop(op_name, sort=None):
-        def func(self, other, sort=sort):
-            self._assert_can_do_setop(other)
-            other = ensure_index(other)
-            if not isinstance(other, IntervalIndex):
-                result = getattr(self.astype(object), op_name)(other)
-                if op_name in ('difference',):
-                    result = result.astype(self.dtype)
-                return result
-            elif self.closed != other.closed:
-                msg = ('can only do set operations between two IntervalIndex '
-                       'objects that are closed on the same side')
-                raise ValueError(msg)
+    @Appender(_index_shared_docs['intersection'])
+    @setop_check(op_name='intersection')
+    def intersection(self, other, sort=False):
+        if self.left.is_unique and self.right.is_unique:
```

Comment:
I _think_ you'll want to check `is_overlapping` instead of `is_unique`, since overlapping is really what should determine the usage of `get_indexer` vs. `get_indexer_non_unique` (`get_indexer` is currently a bit buggy).  It looks like this is causing some incorrect results on your branch:
```python
In [1]: import pandas as pd

In [2]: ii1 = pd.IntervalIndex.from_tuples([(0, 3), (0, 2)])

In [3]: ii2 = pd.IntervalIndex.from_tuples([(0, 2), (1, 3)])

In [4]: ii1
Out[4]: 
IntervalIndex([(0, 3], (0, 2]],
              closed='right',
              dtype='interval[int64]')

In [5]: ii2
Out[5]: 
IntervalIndex([(0, 2], (1, 3]],
              closed='right',
              dtype='interval[int64]')

In [6]: ii1.intersection(ii2)
Out[6]: 
IntervalIndex([(0, 3], (0, 2]],
              closed='right',
              dtype='interval[int64]')
```